### ISSUE TITLE
IOS Problem with Xcode 8, due to HAVE_CLOCK_GETTIME

### ISSUE THREAD

**Issue by mgorges** [gambit/gambit#215]
Hi @feeley, just wanted to let you know about a potential problem with XCode 8. We fixed this in [LambdaNative](http://www.lambdanative.org) earlier today for 4.7.9 (or 4.7.2). Apple has introduced some realtime clock functions in iOS 10 and now returns HAVE_CLOCK_GETTIME as 1. So if you build apps that should run for iOS 9 and below, where the clock functions don't exist, they will compile but then segfault during runtime. The ugly, but working patch, we came up with is to fix this in the configure script, where we comment the `#define HAVE_CLOCK_GETTIME 1` for iOS and things work. You probably should come up with a proper way to do so. Thanks!

**Comment by feeley** [comment]
Thanks for the issue.  I assume this is happening because Gambit is being compiled for iOS outside of Xcode (where the project's settings would indicate the targeted iOS version). Are you using the build-gambit-iOS script? Maybe that can be modified to allow the selection of the appropriate SDK.

**Comment by mgorges** [comment]
Maybe, but I actually think this is a broader issue, as you can only specify the minimum supported iOS version when building. XCode 8 sets HAVE_CLOCK_GETTIME to true for all versions of iOS, which compiles and runs fine on iOS 10 (presumably, don't have one on my desk at the moment). However, for any older iOS versions, including the one we tested with (9.3.5), it compiles fine but than crashes when starting an application because functions are undeclared. So unless you no longer want to support iOS versions before 10, you might need to look into this. - I'll ask @clpetersen to comment as he might be able to explain this better than me.

**Comment by clpetersen** [comment]
I'm not sure if I can explain better, but what we observed is that when gambit is compiled with the iOS SDK cross-compiler in XCode 8, the gambit configure script will detect the presence of clock_gettime, and the resulting gambit library will hence contain references to this function. However, versions of iOS prior to 10 does not define clock_gettime, and the binaries thus segfault when deployed on those older versions of iOS.

**Comment by feeley** [comment]
I understand.  But it is not clear what your build process is.  Are you using the build-gambit-iOS script, or something similar?  It sets a specific SDK (the latest available if I remember correctly).  So maybe it is as simple as setting an older SDK in that script.

**Comment by mgorges** [comment]
We did not use the script but configure gambc manually. We do regular `./configure` with a whole bunch of options.

**Comment by feeley** [comment]
Do you use something like

```CC="xcrun -sdk $sdk_name gcc -isysroot $ios_sdk_dir -arch $arch -miphoneos-version-min=$iphoneos_version_min"```

with an appropriate ```$iphoneos_version_min```?  This is what build-gambit-iOS uses.

**Comment by clpetersen** [comment]
Yes, that's the gist of what we do. The problem here is that this issue cannot be circumvented with `iphoneos-version-min` if you want to build for iOS prior to version 10. You'd basically need the opposite, an "`iphoneos-version-max`", to prevent the inclusion of the new function, and that option doesn't exist (not that it would be the right solution). This is a bug in XCode 8.

**Comment by feeley** [comment]
I see... a bug in XCode 8 because it doesn't respect the iphone-version-min/max range in determining if clock_gettime is available. Bummer! So perhaps an ifdef iOS to disable clock_gettime will do until there is a better way to detect it reliably.

**Comment by feeley** [comment]
I came up with this solution in lib/os.h:

```
/* Make exceptions for iOS which doesn't support some features it claims to. */

#define USE_NONBLOCKING_FILE_IO

#ifdef HAVE_TARGETCONDITIONALS_H
#include <TargetConditionals.h>
#ifdef TARGET_OS_IPHONE
#if TARGET_OS_IPHONE == 1
#undef USE_NONBLOCKING_FILE_IO
#undef HAVE_CLOCK_GETTIME
#endif
#endif
#endif
```

Let me know if you see a problem with this.

**Comment by mgorges** [comment]
@feeley this looks sensible to me. The redundancy of the ifdef and if both checking for TARGET_OS_IPHONE is intentional, correct? Thanks for looking at his so quickly!

**Comment by feeley** [comment]
That's the ifdef sequence I found on another site for detecting iOS. It just works so I won't look into improving it!

**Comment by feeley** [comment]
Fixed by commit eafefcc905619be1b5ac680f5fd5c3d6a20519b4
